map.c

```diff
@@ -409,8 +633,157 @@ typedef struct {
+	mm_time_chain_min = s->buf[0]->timers[MM_TIME_CHAIN];
+	mm_time_chain_max = s->buf[0]->timers[MM_TIME_CHAIN];
+	mm_time_chain_avg = s->buf[0]->timers[MM_TIME_CHAIN];
+	mm_time_align_min = s->buf[0]->timers[MM_TIME_ALIGN];
+	mm_time_align_max = s->buf[0]->timers[MM_TIME_ALIGN];
+	mm_time_align_avg = s->buf[0]->timers[MM_TIME_ALIGN];
+	for (int i = 1; i < p->n_threads; ++i) {
+		mm_time_seed_min = MIN(mm_time_seed_min, s->buf[i]->timers[MM_TIME_SEED]);
+		mm_time_chain_min = MIN(mm_time_chain_min, s->buf[i]->timers[MM_TIME_CHAIN]);
+		mm_time_align_min = MIN(mm_time_align_min, s->buf[i]->timers[MM_TIME_ALIGN]);
+		mm_time_seed_max = MAX(mm_time_seed_max, s->buf[i]->timers[MM_TIME_SEED]);
+		mm_time_chain_max = MAX(mm_time_chain_max, s->buf[i]->timers[MM_TIME_CHAIN]);
+		mm_time_align_max = MAX(mm_time_align_max, s->buf[i]->timers[MM_TIME_ALIGN]);
+		mm_time_seed_avg += s->buf[i]->timers[MM_TIME_SEED];
+		mm_time_chain_avg += s->buf[i]->timers[MM_TIME_CHAIN];
+		mm_time_align_avg += s->buf[i]->timers[MM_TIME_ALIGN];
+	}
+	mm_time_seed_sum = mm_time_seed_avg;
+	mm_time_chain_sum = mm_time_chain_avg;
+	mm_time_align_sum = mm_time_align_avg;
+	mm_time_seed_avg /= p->n_threads;
+	mm_time_chain_avg /= p->n_threads;
+	mm_time_align_avg /= p->n_threads;
+}
+
+
+#if defined(__AMD_SPLIT_KERNELS__)
+static void worker_for(void *_data, long i, int tid) // kt_for() callback
+{
+	step_t *s = (step_t*)_data;
+	int j, iread, off, pe_ori = s->p->opt->pe_ori;
+	double t = 0.0;
+	mm_tbuf_t *b = s->buf[tid];
+	mm_trbuf_t *tr = s->trbuf[tid];
+
+	// Check if this is a valid read
+	if (i != -1) {
+
+		off = s->seg_off[i];
+		iread = tr->count;
+		tr->i[iread] = i;
+		tr->count++;
+
+		assert(s->n_seg[i] <= MM_MAX_SEG);
+		if (mm_dbg_flag & MM_DBG_PRINT_QNAME) {
+			fprintf(stderr, "QR\t%s\t%d\t%d\n", s->seq[off].name, tid, s->seq[off].l_seq);
+			t = realtime();
+		}
+
+		tr->qlens[iread] = (int *) malloc (s->n_seg[i] * sizeof(int));
+		tr->qseqs[iread] = (const char **) malloc (s->n_seg[i] * sizeof(const char *));
+		for (j = 0; j < s->n_seg[i]; ++j) {
+			if (s->n_seg[i] == 2 && ((j == 0 && (pe_ori>>1&1)) || (j == 1 && (pe_ori&1))))
+				mm_revcomp_bseq(&s->seq[off + j]);
+			tr->qlens[iread][j] = s->seq[off + j].l_seq;
+			tr->qseqs[iread][j] = s->seq[off + j].seq;
+		}
+		// Seed
+		if (s->p->opt->flag & MM_F_INDEPEND_SEG) {
+			for (j = 0; j < s->n_seg[i]; ++j) {
+				mm_map_seed(s->p->mi, 1, &tr->qlens[iread][j], &tr->qseqs[iread][j], &s->n_reg[off+j], &s->reg[off+j], b, s->p->opt, s->seq[off+j].name, &tr->rep_len[iread], &tr->qlen_sum[iread], &tr->n_mini_pos[iread], &tr->mini_pos[iread], &tr->n_a[iread], &tr->a[iread], tr->km[iread]);
+			}
+		} else {
+			mm_map_seed(s->p->mi, s->n_seg[i], tr->qlens[iread], tr->qseqs[iread], &s->n_reg[off], &s->reg[off], b, s->p->opt, s->seq[off].name, &tr->rep_len[iread], &tr->qlen_sum[iread], &tr->n_mini_pos[iread], &tr->mini_pos[iread], &tr->n_a[iread], &tr->a[iread], tr->km[iread]);
```

Comment:
As we've discussed, because user's host memory may not be able to hold all intermindiate values, we should copy the seeding result (i.e. anchors) to device right after it is generated and free it on host. But what should we do if the device memory is also full? How to schedule the priority of device memory usage of each worker? Maybe we should have another thread manage all the memcpy and each worker just registers the anchors they've generated?

Reply:
If possible, I still want to try aggregate the seeding results from different workers to run on one kernel launch as soon as the number of anchors reaches our batch size.

Reply:
Let's discuss both points in more detail in our meeting. My initial thoughts after reading your comments: 
1. Maybe for the first iteration, we can keep it simple and see the max N_ACCUM and n_threads we can accomodate in the memory we have. Check if the seeding output (anchors produced for the batch of N_ACCUM reads) is sufficient to saturate the GPU with the chaining kernel.
2. Since each worker thread is accumulating N_ACCUM reads, it may be sufficient to just take the seeding outputs of the whole batch and then process them with the chaining kernel as you see fit. It may be one or multiple kernel launches by pooling the anchors of the various reads in the batch. Again, I like to keep it simple because our end goal is acceptance by the developer community.